### PR TITLE
Make SSL tests work

### LINKED ISSUE #6919: HTTPS tests break in WPT with certification validation error
I assume we need to set up Servo to load certificates from a different location than usual; @jgraham ?

### PR DESCRIPTION
**Pull request by jdm**
These changes fix #6919.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15784)

### CODE REVIEW

**Comment by jdm** [comment]
This addresses the issue tracked in #6919.

**Comment by jdm** [comment]
Whoops, I only meant to include the last commit in this branch. Anyways, ignore the others.

**Comment by jgraham** [comment]
Firefox has a pref that mimics the behaviour of /etc/hosts so from the point of view of the networking code the domain and certificate match [1] and uses certutil to actually install the certificates into the local certificate database [2].

[1] https://dxr.mozilla.org/mozilla-central/source/testing/web-platform/harness/wptrunner/browsers/firefox.py#139
[2] https://dxr.mozilla.org/mozilla-central/source/testing/web-platform/harness/wptrunner/browsers/firefox.py#227

**Comment by jdm** [comment]
@nox, the network changes should be familiar to you now. @jgraham could you sign off on the wptrunner changes?

File: components/net/connector.rs [hunk 1]
```diff
@@ -7,15 +7,19 @@ use hyper::net::HttpsConnector;
...
 pub fn create_ssl_client(certificate_file: &str) -> OpensslClient {
-    let ca_file = &resources_dir_path()
-        .expect("Need certificate file to make network requests")
-        .join(certificate_file);
+    let ca_file = match opts::get().certificate_path {
+        Some(ref path) => PathBuf::from(path),
+        None => resources_dir_path()
+            .expect("Need certificate file to make network requests")
+            .join(certificate_file),
+    };
```
Comment (nox) [inline components/net/connector.rs]: Wait, so the argument is ignored if there is a `certificate_path`? Please either pass the `certificate_path` option to the function in callers, or remove the argument altogether; it feels weird that the option takes precedence over the argument.

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -129,7 +128,7 @@ impl NetworkConnector for NetworkHttpRequestFactory {
     type Stream = PooledStream<HttpsStream<SslStream<HttpStream>>>;
 
     fn connect(&self, host: &str, port: u16, scheme: &str) -> Result<Self::Stream, HttpError> {
-        self.connector.connect(&replace_host(host), port, scheme)
+        self.connector.connect(host, port, scheme)
```
Comment (nox) [inline components/net/http_loader.rs]: <s>This implementation exists only to do that `replace_host` call, now that it is gone you can pass `&self.connector` again to `HyperRequest::with_connector` in `NetworkHttpRequestFactory::create` and remove this impl altogether.</s>

Hah, you killed the whole wrapper in next commit, nice!

File: components/net/http_loader.rs [hunk 3]
```diff
@@ -1103,7 +1079,7 @@ fn http_network_fetch(request: &Request,
     // do not. Once we support other kinds of fetches we'll need to be more fine grained here
     // since things like image fetches are classified differently by devtools
     let is_xhr = request.destination == Destination::None;
-    let wrapped_response = obtain_response(&factory, &url, &request.method,
+    let wrapped_response = obtain_response(context.connector.clone(), &url, &request.method,
```
Comment (nox) [inline components/net/http_loader.rs]: AFAIK this clone is pointless, and you can pass `&Pool<Connector>` to `obtain_response`.

**Review by nox — CHANGES_REQUESTED** [review]
Commits "Update mozilla/ WPT results." and "Update results for remaining https tests in WPT." need to be squashed with "Only replace hosts when making TCP connections, not when verifying SSL certificates."

And you would please the old man I am if you would shorten that last commit title I had trouble pasting.

File: components/net/http_loader.rs (hunk 3 above)
Comment (jdm) [inline components/net/http_loader.rs]: That was what I tried originally, and Rust's type checker became _so confused_ that I didn't want to try to debug. Please don't make me investigate it; it ends up trying to use the closure connector feature instead and it's not worth it.

**Comment by jgraham** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 942 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15784)*

**Comment by jgraham** [comment]
wptrunner changes look fine; please remember to upstream them.

**Pull request merged**